test/test_latest_time_policy.cpp

```diff
@@ -0,0 +1,277 @@
+                                                     std::placeholders::_2,
+                                                     std::placeholders::_3));
+    p.reserve(12U);
+    q.reserve(6U);
+    r.reserve(3U);
+    for(std::size_t idx = 0U; idx < 12U; ++idx)
+    {
+      MsgPtr p_idx(std::make_shared<Msg>()); p_idx->data = idx; p.push_back(p_idx);
+      if(idx % 2U == 0U)
+      {
+        MsgPtr q_idx(std::make_shared<Msg>()); q_idx->data = idx; q.push_back(q_idx);
+      }
+      if(idx % 4U == 0U)
+      {
+        MsgPtr r_idx(std::make_shared<Msg>()); r_idx->data = idx; r.push_back(r_idx);
+      }
+    }
+  }
+};
+
+
+TEST_F(LatestTimePolicy, Leading)
+{
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    if(idx % 2U == 0U)
+    {
+      sync.add<1>(q[idx / 2U]);
+    }
+    if(idx % 4U == 0U)
+    {
+      sync.add<2>(r[idx / 4U]);
+    }
+    sync.add<0>(p[idx]);
+
+    EXPECT_EQ(h.count_, idx);
+    if(idx > 0)
+    {
+      EXPECT_EQ(h.p_->data, p[idx]->data);
+      EXPECT_EQ(h.q_->data, q[idx / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[idx / 4U]->data);
+    }
+    else
+    {
+      EXPECT_FALSE(h.p_);
+      EXPECT_FALSE(h.q_);
+      EXPECT_FALSE(h.r_);
+    }
+
+    rate.sleep();
+  }
+}
+
+TEST_F(LatestTimePolicy, Trailing)
+{
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 8U; ++idx)
+  {
+    if(idx % 2U == 1U)
+    {
+      sync.add<1>(q[(idx - 1U) / 2U]);
+    }
+    if(idx % 4U == 3U)
+    {
+      sync.add<2>(r[(idx - 3U) / 4U]);
+    }
+    sync.add<0>(p[idx]);
+
+    if (idx > 2U)
+    {
+      EXPECT_EQ(h.count_, idx - 2U);
+      EXPECT_EQ(h.p_->data, p[idx]->data);
+      EXPECT_EQ(h.q_->data, q[(idx - 1U) / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[(idx - 3U) / 4U]->data);
+    }
+    else
+    {
+      EXPECT_FALSE(h.p_);
+      EXPECT_FALSE(h.q_);
+      EXPECT_FALSE(h.r_);
+    }
+
+    rate.sleep();
+  }
+}
+
+TEST_F(LatestTimePolicy, ChangeRate)
+{
+  rclcpp::Rate rate(50.0);
+  for(std::size_t idx = 0U; idx < 12U; ++idx)
+  {
+    if(idx % 2U == 1U)
+    {
+      sync.add<1>(q[(idx - 1U) / 2U]);
+    }
+
+    if(idx % 4U == 3U)
+    {
+      sync.add<2>(r[(idx - 3U) / 4U]);
+    }
+
+    if (idx < 4U)
+    {
+      sync.add<0>(p[idx]);
+    }
+    else  // Change rate of p (still 50Hz @ idx == 4)
+    {
+      if(idx % 3U == 1U)
+      {
+        static std::size_t p_idx = 3U;
+        sync.add<0>(p[++p_idx]);
+      }
+    }
+
+    // operates like "Trailing" test for idx <= 3
+    if (idx == 3U)
+    {
+      EXPECT_EQ(h.count_, idx - 2U);
+      EXPECT_EQ(h.p_->data, p[idx]->data);
+      EXPECT_EQ(h.q_->data, q[(idx - 1U) / 2U]->data);
+      EXPECT_EQ(h.r_->data, r[(idx - 3U) / 4U]->data);
+    }
+    // rate of p still 50Hz @ idx==4, then change rate of p lower than q
+    // will not publish again until idx==9
+    // since rate of q is 25Hz, p was 50Hz @ idx==4,
+    // and new rate of p computed after q received @ idx==7
```

Comment:
I'm a bit confused about this test case. If I understand correctly, all 3 (p, q and r) have new values at `idx==7`. But that's not being published until `idx==9`. 

I understood that the filter was supposed to publish at the fastest rate possible, whenever there's a new message. So I'd expect it to publish at `idx==5` (when q changes), and again at `idx==7` (when they all change), so the subscriber always has the most up-to-date information.

Is my assumption correct?

Reply:
I could try to change the behavior, but the last computed rate for `p` is 50Hz until it is received again when it is received at `idx==7` but it is received after `q` at `idx==7`. `q` would be the new pivot, but the policy doesn't know that yet until `q` is received at `idx==9`. I could change this to detect that `p` is late, but I'd have to think about side-effects and what to do if it's late.

Reply:
> I'd expect it to publish at `idx==5` (when q changes)

The computed rate for `p` is still 50Hz at `idx==5` since it was computed as 50Hz at `idx==4` and won't be computed again until `idx==7`.

Reply:
I now loop over the sorted rates and pick the first message that isn't late as the pivot. This has the effect of publishing on `idx==7`

Reply:
Basically, q now detects that p is late at idx==7, but can't at idx==5 since p still has a chance to arrive after q and still be on time.

Reply:
> publish at the fastest rate possible, whenever there's a new message

This would have the effect of publishing at a fairly erratic rate. For example, if `p` and `q` were both 50Hz, we'd get two publishes of all messages. If `p` and `q` came in 0.01s apart, everything would look like a ZOH @ 100Hz. If they came in closer/further apart, the publishing would be galloping. The publishing would be much more erratic for more messages.

Reply:
Thanks for the explanation, the consequences of publishing all the time are clear to me now 👍🏽